deep_qa/training/trainer.py

```diff
@@ -534,11 +622,9 @@ def _get_callbacks(self):
         # Some witchcraft is happening here - we don't specify the epoch replacement variable
         # checkpointing string, because Keras does that within the callback if we specify it here.
         if self.save_models:
-
-            checkpoint_callback = ReplicaModelCheckpoint if self.num_gpus > 1 else ModelCheckpoint
-            checkpointing = checkpoint_callback(self.model_prefix + "_weights_epoch={epoch:d}.h5",
-                                                save_best_only=True, save_weights_only=True,
-                                                monitor=self.validation_metric)
+            checkpointing = ModelCheckpoint(self.model_prefix + "_weights_epoch={epoch:d}.h5",
```

Comment:
Oh, I get it.  I missed the `primary_model = tower_models[0]` line, and was thinking the `primary_model` was some global thing that contained the tower models.  But that was how the previous code works, and this one just hacks things a bit, overwriting the first tower model's training function with the losses and gradients of all of the tower models.

And this only happens at training time, and not a validation time, because we're only overwriting the train function?

In any event, the information in this discussion should be summarized in the docstring to this method, I think.